### PR TITLE
ai uploads now have a gps signal

### PR DESCRIPTION
**Pull request by Tlaltecuhtli**
:cl:
balance: ai/borg uploads now have a gps signal when built which will be active until its monitor gets decostructed
balance: ai/borg uploads take slighty longer to deconstruct
/:cl:

maint uploads are a pain in the ass to deal for all the parties involved as you ll have x and y uploading laws every second to counter the other while the ai switches from asimov and rogue back and forward and suicides

### CODE REVIEW

**Comment by HomocidlePancakes** [comment]
Holy shit. PR of the year right here.

**Comment by anconfuzedrock** [comment]
MWAHAHA MORE GPS SIGNALS FIRST THE UPLOAD NEXT YO MOMMA’S HOUSEZ

**Comment by HomocidlePancakes** [comment]
Do AI boards have a GPS signal or consoles?

**Comment by Tlaltecuhtli** [comment]
read the OP

**Comment by ondrej008** [comment]
@Tlaltecuhtli can the AI get the GPS signal? I know only that Mining module cyborgs can use a GPS, so this would probably only help the crew.

**Comment by ExcessiveUseOfCobblestone** [comment]
It'll always be on the same Z level so you could just display a message to the AI stating the coords.

That feels a lot safer than a miner/literally anyone looking at the GPS and noticing this signal popping up and instantly calling it out.

**Comment by zxaber** [comment]
I still say that carded AIs shouldn't be eligible for law changes. But this is a good alternative fix.

**Comment by Ratvar** [comment]
Can you add this behavior to comms console as well?

**Comment by JMoldy** [comment]
Merge this yesterday.

**Review by ExcessiveUseOfCobblestone — CHANGES_REQUESTED** [review]
default the GPS to `tracking = FALSE`, only activating once it's actually been used (see hit by law board).

I would also prefer the default computer to not have this.

**Comment by Dorsisdwarf** [comment]
Wait, its only active when the monitor is connected? cant you just screwdriver -> upload law -> screwdriver

**Comment by zxaber** [comment]
That's why the decon takes longer now, presumably.

Far as I can tell, this isn't a nerf to tators subverting the AI, it's meant to be a way to stop law spammers who are doing nothing but making the AI's life miserable.

**Comment by Tlaltecuhtli** [comment]
@ExcessiveUseOfCobblestone what do you mean
@Ratvar comms console already have the message thing printed on all consoles
@ExcessiveUseOfCobblestone ai being aware of uploads may cause them to premtively suicide because law 1

**Comment by carshalash** [comment]
Oh thank god.

**Comment by AnonymousTGplayer** [comment]
I fail to see the initial issue as I feel this PR oozes plenty of salt. Sure, maintenance uploads are a pain to deal with, but that's kind of the point of them. I feel this PR punishes traitors (specifically non-science/engineering traitors)/ops with hacked AI boards rather than players who law spam for fun.

> Far as I can tell, this isn't a nerf to tators subverting the AI, it's meant to be a way to stop law spammers who are doing nothing but making the AI's life miserable.
Spamming (law/chat or otherwise) is precedent enough to violate rule 1 (and getting bwoinked if the admins have had enough regardless if you're traitor or not) and law spam usually result in admins (especially if you ahelp regarding just this) deciding that the AI gets a free malf due to "software bugs".
I feel a cooldown on the boards or upload would be better fix to deal with spammers than adding a magic GPS signal to every upload.

**Comment by zxaber** [comment]
Admins aren't always available to grant malf or handle the upload spam. A cooldown would prevent legitimate back-to-back uploads (such as a set of custom laws), though a built-up cooldown or temporary "burnout" of the board similar to flashes might not be a bad idea.

Really, though, this isn't going to hurt tators unless people already know they’re gonna upload a law. No one will be watching for upload locations normally, and there's no alert the AI gets an upload. Under normal circumstances, by the time anyone realizes the upload happened, the traitor will likely not be near the location where they uploaded anymore.

**Comment by Cyprex** [comment]
>comms console already have the message thing printed on all consoles
Yes, but that only prints the area name, and unreliably, and there are some ways around that. Comms consoles should also have this behaviour, the only issue I can see is GPS spam, which could be mitigated by adding them into a seperate tab on the GPS.

**Comment by SouDescolado** [comment]
Btw, you used it all correctly, but you named it time to scewdrive, when I think you meant screwdrive.

**Comment by ExcessiveUseOfCobblestone** [comment]
I don't want there to be any indication a non-default (mapload) upload exists UNLESS it has already been used to upload a law.

I don't really want there to be a signal for the default upload either, it's useless.

**Comment by MacHac** [comment]
If miners have their GPSes up, and they usually do while hunting goodies, they'll see the signal pop up on the station for a moment.  Even if they don't know where the console is, they can yell about it over comms, and the AI core is now being raided by all of security + every bored validhunter on the station.

Alternate suggestion: When the AI's laws are changed, it gets a message saying where the change came from.  The location of the most recent change also appears in the carded AI's interface.  

If the AI's being one-off subverted, its laws almost certainly forbid it from revealing the location; If the AI's being lawspammed, anyone that cards it will be able to find the console without the console user realizing.

**Comment by Cyprex** [comment]
Having a signal on the default upload would help players realize this change.

**Comment by ExcessiveUseOfCobblestone** [comment]
We have a changelog if players want to stay in the know, we don't need a useless entry autoupdating.

> If miners have their GPSes up, and they usually do while hunting goodies, they'll see the signal pop up on the station for a moment. Even if they don't know where the console is, they can yell about it over comms, and the AI core is now being raided by all of security + every bored validhunter on the station.
> 
> Alternate suggestion: When the AI's laws are changed, it gets a message saying where the change came from. The location of the most recent change also appears in the carded AI's interface.
> 
> If the AI's being one-off subverted, its laws almost certainly forbid it from revealing the location; If the AI's being lawspammed, anyone that cards it will be able to find the console without the console user realizing.

Yes I'm afraid of that as well, but i'm not sure how to fix "I want to be able to know exactly where a console is so I can sucker punch it... but not too much"

**Comment by Tlaltecuhtli** [comment]
@ExcessiveUseOfCobblestone having the tracking turn on only when laws is uploaded and then turn off is worse tho i think its better as currently is

**Comment by 4dplanner** [comment]
You can just leave the console almost constructed anyway

**Comment by ExcessiveUseOfCobblestone** [comment]
> @ExcessiveUseOfCobblestone having the tracking turn on only when laws is uploaded and then turn off is worse tho i think its better as currently is

It's not meant to turn off after use, it's meant to START off and turn on indefinitely after use. Like I said if you're worried about deconstruction/reconstruction put it on the circuitboard instead.

**Comment by zxaber** [comment]
Would it be better if, instead of emitting a GPS signal, all AI uploads simply listed the coordinates where the most recent law change happened, regardless if the console it actually still there? It would solve the worry that random miners or whoever could call out traitors, but would mean that someone deconning the maint upload console could still be found.

**Comment by ExcessiveUseOfCobblestone** [comment]
It doesn't use a UI unless you want it to "say" it in which case it would be useless unless you stood by one.

Putting the GPS on the circuitboard would fix the issue of deconning to bypass the mechanic.

**Comment by zxaber** [comment]
Maybe, but is the point of this to dissuade against maintenance law spam, or give a counter to proper traitors subverting the AI? I was under the impression that we were trying to not touch normal traitors, but having a GPS link to their pocket once the subverting is done is a pretty big nerf to the subversion mechanic. At the very least, even if they drop off the board somewhere, it's an unintuitive way for their work to be found out if someone just happens to glance at a GPS.

I thought we were limiting this to be seen only from other AI uploads. Having GPS pinpointer that anyone might notice is pretty terrible, and having it be permanently on the board would be even worse.

**Comment by ExcessiveUseOfCobblestone** [comment]
> Above Post

You can't exactly differentiate the two is the main issue. As I said:

>i'm not sure how to fix "I want to be able to know exactly where a console is so I can sucker punch it... but not too much"

If the only metric for this PR is that it stops spam at the cost of previous gameplay interactions then it may as well be closed, we have admins for those abusing mechanics.

**Comment by zxaber** [comment]
Punishing it by making it so easy to powergame against seems not great.

**Comment by ExcessiveUseOfCobblestone** [comment]
> Punishing it by making it so easy to powergame against seems not great.

Right hm....

I think I should go back to my original suggestion of printing the coords to the AI.

File: code/game/machinery/computer/law.dm [hunk 1]
```diff
@@ -3,6 +3,23 @@
 /obj/machinery/computer/upload
 	var/mob/living/silicon/current = null //The target of future law uploads
 	icon_screen = "command"
+	var/obj/item/gps/internal
```
Comment (ShizCalev) [inline code/game/machinery/computer/law.dm]: mind renaming this to internal_gps?

**Pull request merged**